src/nvim/api/vim.c

```diff
@@ -199,6 +199,46 @@ Integer nvim_get_hl_id_by_name(String name)
   return syn_check_group((const char_u *)name.data, (int)name.size);
 }
 
+void nvim_set_theme_provider(Integer ns_id, LuaRef cb, Error *err)
+  FUNC_API_SINCE(7)
+  FUNC_API_LUA_ONLY
+{
+  hl_theme_set_provider((int)ns_id, api_new_luaref(cb));
+  // TODO: more lazy:
+  if (active_theme == ns_id) {
+    redraw_later(NOT_VALID);
+  }
+}
+
+/// TODO: ns_id = 0, should modify :highlight namespace
+/// TODO val should take update vs reset flag
+void nvim_set_highlight(Integer ns_id, String name, Dictionary val, Error *err)
+  FUNC_API_SINCE(7)
+{
+  int hl_id = syn_check_group( (char_u *)(name.data), (int)name.size);
+  HlAttrs attrs = dict2hlattrs(val, true);
+  hl_theme_def((int)ns_id, hl_id, attrs);
+}
+
+void nvim_set_theme(Integer ns_id, Boolean immediate, Error *err)
+  FUNC_API_SINCE(7)
+  FUNC_API_FAST
+{
+  if (ns_id >= 0) {
+    active_theme = (int)ns_id;
+  }
+
+  highlight_changed();
+  // TODO(bfredl): be "smart" and assume "immediate" from updating_screen ?
+  if (immediate) {
+    if (updating_screen && drawn_win != NULL) {
```

Comment:
I am not a fan of adding another global drawn_win, redrawing everything should be fine or maybe pass a  `nvim_win_set_theme`

Reply:
it was a hack to enable an ongoing design process. This will be rebased on top of luahl after which no more globals should be needed.